### ISSUE TITLE
Failed to set EFI variable HibernateLocation: Invalid argument - when hibernating

### ISSUE THREAD

**Issue by linuxandos2** [systemd/systemd#32331]
### systemd version the issue has been seen with

sys-apps/systemd-9999

### Used distribution

gentoo updated to today

### Linux kernel version used

sys-kernel/gentoo-sources-6.8.3 on amd64

### CPU architectures issue was seen on

other

### Component

systemd

### Expected behaviour you didn't see

system fails to hibernate

### Unexpected behaviour you saw

system should hibernate

### Steps to reproduce the problem

systemctl hibernate

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
journalctl:
Apr 18 10:51:13 localhost systemd-logind[1861]: The system will hibernate now!
Apr 18 10:51:15 localhost systemd[1]: Reached target sleep.target.
Apr 18 10:51:15 localhost systemd[1]: Starting systemd-hibernate.service...
Apr 18 10:51:15 localhost systemd-sleep[5690]: Failed to set EFI variable HibernateLocation: Invalid argument
Apr 18 10:51:15 localhost systemd[1]: systemd-hibernate.service: Main process exited, code=exited, status=1/FAILURE
Apr 18 10:51:15 localhost systemd[1]: systemd-hibernate.service: Failed with result 'exit-code'.
Apr 18 10:51:15 localhost systemd[1]: Failed to start systemd-hibernate.service.
Apr 18 10:51:15 localhost systemd[1]: Dependency failed for hibernate.target.
Apr 18 10:51:15 localhost systemd[1]: hibernate.target: Job hibernate.target/start failed with result 'dependency'.
Apr 18 10:51:15 localhost systemd-logind[1861]: Operation 'sleep' finished.
Apr 18 10:51:15 localhost systemd[1]: Stopped target sleep.target.
```

**Comment by poettering** [comment]
> ###
> 
> sys-apps/systemd-9999

I have no idea what that means, but that's not an existing systemd version.

**Comment by poettering** [comment]
> ### CPU architectures issue was seen on
> 
> other

uh?

i never heard of that cpu architecture. Is EFI even supported on that arch?

**Comment by linuxandos2** [comment]
version of sytemd: systemd 255 (255)
sys-apps/systemd-9999 in gentoo means the latest version from github
sorry for the architecture other it is x86_64

**Comment by linuxandos2** [comment]
restored backup of 5 days earlier and reapplied all upgrades.
problem disappeared
none of the packages upgrading had anything to do with systemd or kernel

**Comment by linuxandos2** [comment]
After doing some ugrades  on the restored system like:

net-misc/curl-8.7.1-r2 
net-libs/nodejs-20.12.1
mail-client/thunderbird-115.10.1
www-client/chromium-124.0.6367.60
the same problem appears again.

I really do not see any relation with that problem

Also did following just to exclude:
 removed the ubuntu partitions on same disk.
did restore of the efi partition
did factory restore of the firmware
ran memtest for 8 hours
did grub-mkconfig
no succes.

you can see the my topic on the gentoo forum:
https://forums.gentoo.org/viewtopic-t-1168514.html

**Comment by YHNdnzj** [comment]
Hmm, I'd really guess that the firmware is the culprit here. Maybe fiddling with files in the ESP changed the boot entries in NVRAM, so that it exposes such unstable behavior?

Does other efivar-based functionalities work for you? E.g. systemctl reboot --boot-loader-entry=?

**Comment by poettering** [comment]
anything in the kernel logs about this?

if efivarfs refuses our writes this looks very much like a firmware issue

**Comment by logrusx** [comment]
From the original thread:

```Apr 18 09:26:47 localhost systemd-logind[1855]: The system will hibernate now!
Apr 18 09:26:48 localhost systemd[1]: Reached target Sleep.
Apr 18 09:26:48 localhost systemd[1]: Starting System Hibernate...
Apr 18 09:26:48 localhost systemd-sleep[47832]: Failed to set EFI variable HibernateLocation: Invalid argument
Apr 18 09:26:48 localhost systemd[1]: systemd-hibernate.service: Main process exited, code=exited, status=1/FAILURE
Apr 18 09:26:48 localhost systemd[1]: systemd-hibernate.service: Failed with result 'exit-code'.
Apr 18 09:26:48 localhost systemd[1]: Failed to start System Hibernate.
Apr 18 09:26:48 localhost systemd[1]: Dependency failed for System Hibernation.
Apr 18 09:26:48 localhost systemd[1]: hibernate.target: Job hibernate.target/start failed with result 'dependency'.
Apr 18 09:26:48 localhost systemd[1]: Stopped target Sleep.
Apr 18 09:26:48 localhost systemd-logind[1855]: Operation 'hibernate' finished```

Nothing related in dmesg: https://bpa.st/QA3A

**Comment by logrusx** [comment]
> Does other efivar-based functionalities work for you? E.g. systemctl reboot --boot-loader-entry=?

I think that only works with systemd's bootloader, so if OP uses another solution only --firmware-setup option may be tested that way if it's relevant.

**Comment by linuxandos2** [comment]
systemctl reboot --boot-loader-entry=help
No boot loader entries discovered.

logrusx pointed me to te solution.

the resume kernel cmdline parameter used to be present, but strangely disappeared.
added resume=UUID=b4b1cbec-4e1f-4da7-88d7-5ebc50492090

now the "systemctl hibernate" works.
the resume is also ok.

thanks a lot.

**Comment by logrusx** [comment]
Although the issue is now closed, I'll add a comment just for the record: if possible, it would be better if the error message is more informative. It's very hard to figure out what's going on now.

**Comment by YHNdnzj** [comment]
> It's very hard to figure out what's going on now.

 I'm pretty sure this is a firmware issue. Also, writing to efivarfs returns EINVAL, and we log about the exact error. I don't know what you mean by "informative"

**Comment by logrusx** [comment]
> I don't know what you mean by "more informative"?

"Invalid argument" is not informative. If there's more information, please expose it.

**Comment by linuxandos2** [comment]
I you want to further investigate the problem I'l be happy to work further whit you

**Comment by poettering** [comment]
> >  I don't know what you mean by "more informative"?
> "Invalid argument" is not informative. If there's more information, please expose it.

We write EFI vars via the kernel's "efivarfs" interface. If an attempt to write there fails with the EINVAL error then you see the log message you are seeing. But the thing is that EINVAL is all we get from the kernel, we have no further information.

Yeah, sucks, but systemd can't show you more info, because we have nothing.

Check kernel logs (i.e. dmesg), maybe there's more. But it might be quite likely that the firmware just returns a generic error too that the kernel just propagates to us.

Hence, check dmesg, maybe update your firmware, i have no better suggestions to make.

**Comment by logrusx** [comment]
@poettering, it's not my problem and I can't do much to investigate it further, but thank you very much for the clarification!

**Comment by midenok** [comment]
Did you fix that? Why it was closed? Still reproducible with 259.5

In my case resume parameter is not missing:

```bash
root@lian:midenok# cat /proc/cmdline
BOOT_IMAGE=/boot/vmlinuz-6.6.140-0606140-generic root=/dev/mapper/vg10-root ro resume=UUID=4e501b94-97e8-43fb-a95f-0cee515236d6 no_console_suspend=1 nomodeset security=none selinux=0 enforcing=0 apparmor=0 capability.disable=1 fbcon=rotate:3 net.ifnames=0 intremap=off rootdelay=3
root@lian:midenok# blkid|grep 4e501b94-97e8-43fb-a95f-0cee515236d6
/dev/nvme0n1p7: LABEL="fury-swap" UUID="4e501b94-97e8-43fb-a95f-0cee515236d6" TYPE="swap" PARTLABEL="swap" PARTUUID="263b5bb2-da1a-4632-9a00-04d8c40a336a"
```

Is writing EFI absolutely neccessary? AFAIU resume= arg should be enough
